# HG changeset patch
# Parent c0e9a9b531bf25a8c95ce9ec25a556b4a724e00a
diff -r c0e9a9b531bf -r de5b5da27f1b model/dce-credentials.cc
--- a/model/dce-credentials.cc	Mon Dec 10 13:00:36 2012 +0900
+++ b/model/dce-credentials.cc	Mon Dec 10 19:38:18 2012 +0900
@@ -53,3 +53,15 @@
   int ret = eaccess (rPath.c_str (), mode);
   return ret;
 }
+int dce_chown(const char *path, uid_t owner, gid_t group)
+{
+  // XXX
+  return 0;
+}
+
+int dce_initgroups(const char *user, gid_t group)
+{
+  // XXX
+  return 0;
+}
+
diff -r c0e9a9b531bf -r de5b5da27f1b model/dce-stdio.cc
--- a/model/dce-stdio.cc	Mon Dec 10 13:00:36 2012 +0900
+++ b/model/dce-stdio.cc	Mon Dec 10 19:38:18 2012 +0900
@@ -749,6 +749,16 @@
   va_end (ap);
   return retval;
 }
+int dce___vsnprintf_chk (char *__restrict __s, size_t __n, int __flag,
+						size_t __slen,
+						__const char *__restrict __format, _G_va_list __ap)
+{
+  NS_LOG_FUNCTION (Current () << UtilsGetNodeId ());
+  NS_ASSERT (Current () != 0);
+
+  int retval = vsnprintf (__s, __n, __format, __ap);
+  return retval;
+}
 void dce___fpurge (FILE *stream)
 {
   NS_LOG_FUNCTION (Current () << UtilsGetNodeId () << stream);
diff -r c0e9a9b531bf -r de5b5da27f1b model/dce-stdio.h
--- a/model/dce-stdio.h	Mon Dec 10 13:00:36 2012 +0900
+++ b/model/dce-stdio.h	Mon Dec 10 19:38:18 2012 +0900
@@ -64,6 +64,9 @@
 					   __const char *__restrict __format, ...);
 int dce___snprintf_chk (char *__restrict __s, size_t __n, int __flag,
 						size_t __slen, __const char *__restrict __format, ...);
+int dce___vsnprintf_chk (char *__restrict __s, size_t __n, int __flag,
+						size_t __slen,
+                         __const char *__restrict __format, _G_va_list __ap);
 
 void dce___fpurge (FILE *stream);
 size_t dce___fpending (FILE *stream);
diff -r c0e9a9b531bf -r de5b5da27f1b model/dce-syslog.cc
--- a/model/dce-syslog.cc	Mon Dec 10 13:00:36 2012 +0900
+++ b/model/dce-syslog.cc	Mon Dec 10 19:38:18 2012 +0900
@@ -66,4 +66,5 @@
   Process *process = Current ()->process;
 
   vfprintf (process->syslog, message, args);
+  fprintf (process->syslog, "\n");
 }
diff -r c0e9a9b531bf -r de5b5da27f1b model/dce-unistd.h
--- a/model/dce-unistd.h	Mon Dec 10 13:00:36 2012 +0900
+++ b/model/dce-unistd.h	Mon Dec 10 19:38:18 2012 +0900
@@ -76,6 +76,8 @@
 
 ssize_t dce_pread (int fd, void *buf, size_t count, off_t offset);
 ssize_t dce_pwrite (int fd, const void *buf, size_t count, off_t offset);
+int dce_chown(const char *path, uid_t owner, gid_t group);
+int dce_initgroups(const char *user, gid_t group);
 
 int dce_daemon (int nochdir, int noclose);
 
diff -r c0e9a9b531bf -r de5b5da27f1b model/libc-dce.cc
--- a/model/libc-dce.cc	Mon Dec 10 13:00:36 2012 +0900
+++ b/model/libc-dce.cc	Mon Dec 10 19:38:18 2012 +0900
@@ -96,6 +96,7 @@
 #include <langinfo.h>
 #include <sys/vfs.h>
 #include <termio.h>
+#include <assert.h>
 
 extern void __cxa_finalize (void *d);
 extern int __cxa_atexit (void (*func) (void *), void *arg, void *d);
@@ -137,9 +138,11 @@
 extern int __obstack_vprintf_chk (struct obstack *, int, const char *,
                   _G_va_list) __THROW;
 
+
 typedef void (*func_t) (...);
 
 extern "C" {
+extern void *__memcpy_chk(void *dest, const void *src, size_t n, size_t __destlen);
 
 void libc_dce (struct Libc **libc)
 {
diff -r c0e9a9b531bf -r de5b5da27f1b model/libc-ns3.h
--- a/model/libc-ns3.h	Mon Dec 10 13:00:36 2012 +0900
+++ b/model/libc-ns3.h	Mon Dec 10 19:38:18 2012 +0900
@@ -109,6 +109,7 @@
 NATIVE (strcoll)
 NATIVE (memset)
 NATIVE (memcpy)
+NATIVE (__memcpy_chk)
 NATIVE (bcopy)
 NATIVE (memcmp)
 NATIVE (memmove)
@@ -246,6 +247,9 @@
 DCE (daemon)
 DCE (alarm)
 DCE (readlink)
+DCE (chown)
+DCE (initgroups)
+NATIVE (access)
 
 // SYS/UIO.H
 DCE (readv)
@@ -259,8 +263,8 @@
 DCE (setlinebuf)
 DCE (fseek)
 DCE (ftell)
-DCE (fseeko)
-DCE (ftello)
+//DCE (fseeko)
+//DCE (ftello)
 DCE (rewind)
 DCE (fgetpos)
 DCE (fsetpos)
@@ -301,6 +305,8 @@
 DCE (remove)
 //NATIVE (sscanf)
 NATIVE_WITH_ALIAS2 (sscanf, __isoc99_sscanf)
+NATIVE (flockfile)
+NATIVE (funlockfile)
 
 // STDARG.H
 DCE (vprintf)
@@ -407,6 +413,11 @@
 NATIVE (pthread_setcancelstate)
 NATIVE (pthread_sigmask)
 NATIVE (pthread_equal)
+NATIVE (pthread_spin_init)
+NATIVE (pthread_spin_lock)
+NATIVE (pthread_spin_unlock)
+NATIVE (pthread_spin_destroy)
+
 
 // SEMAPHORE.H
 DCE (sem_init)
@@ -502,6 +513,8 @@
 NATIVE (bindtextdomain)
 NATIVE (textdomain)
 NATIVE (gettext)
+NATIVE (catopen)
+NATIVE (catgets)
 
 // PWD.H
 NATIVE (getpwnam)
@@ -563,6 +576,7 @@
 DCE    (__fprintf_chk)
 DCE    (__snprintf_chk)
 DCE    (__errno_location)
+DCE    (__vsnprintf_chk)
 
 DCE    (__xstat)
 DCE    (__lxstat)
